### PR TITLE
Appclient

### PR DESCRIPTION
**Pull request by red-0ne**
# Additions
* A daemon that listens to native JsonRPC/Websocket clients, signs the requests and forwards them to a `Relayer`
* A `Relayer` endpoint strategy selector interface and naive implementation that picks form the session's `Relayer`s available endpoints
* A `Signer` interface and `Application` implementation that signs relay requests.
* Rewrite part of the `Relayer` proxy to handle Pocket `RelayRequest/Response`s instead of service messages
* A blocks query client extracted from the `Relayer`s client (stripped from signing and tx logic)
* A MakeFile entry to run the daemon

# Changes
* Factoring code used by both the `Relayer` and `AppClient`
* Observable now takes a context for cancellation, simplifying its useage

# TODO
* Use the auth module to fetch pubkeys for verification. The `QueryClient` is added but needs to be used where session changes are emitted
* Websockets going living across multiple sessions do not work half/most of the time. Suspecting the `Relayer`'s connection closing not handled properly
* Signature verification on both `Relayer` and `AppClient` as they need pubkeys fetched by the auth `QueryClient`

----

Reviewers should start by looking into the `smartclient/relayhandler`
1. `relayhandler.go`
2. `httphandler.go`
3. `session_connection.go`
4. `wshandler.go`

Naming is aweful.

## Run
Launch
```sh
make localnet_up
make appclient_start
```
Send requests
```
make cast_relay
```
or
```
make ws_subscribe
```

Query claims
```
make claims_query
```

### CODE REVIEW

File: smartclient/relayhandler/relayhandler.go
Comment (Olshansk) [inline smartclient/relayhandler/relayhandler.go:143]: I think the relay handler's session management business logic should be abstracted out into a `SessionCache`. Wdyt?

**Review by Olshansk — APPROVED** [review]
@red-0ne My 🧠 is fried trying to understand this PR. Here are my key takeaways:

**1. This is so much fucking amazing work. I'm genuinely impressed!**
2. This is pretty much the foundation for the entire business logic of the `gateway`
3. I'm approving this, but when we go post-alpha, we just need to follow some of the engineering practices that I'm 100% sure you're all so familiar with:
- Smaller PRs (e.g. SessionCache as an independent thing)
- Naming. We REALLY need to improve naming. (e.g. this is a gateway, not a smart client)
- Code organization (e.g. add interface for all major structures and don't have a central interfaces file)
- Etc...

Going to approve because I don't want to block anything, but this amazing. It's just a lot of mental overhead to fully understand right now.

File: relayer/client/txs.go
Comment (red-0ne) [reply relayer/client/txs.go:64]: This is now handled at the subscriber level. We pass `ctx` so it handles it on its own instead of setting up the context termination each time.

It is definitely not related to this feature.

File: relayer/miner/miner.go
Comment (red-0ne) [reply relayer/miner/miner.go:58]: This is managed by the subscription itself that gets the `ctx`

File: relayer/proxy/http.go
Comment (red-0ne) [reply relayer/proxy/http.go:28]: `serviceForwardingAddress` is the host:port address to bind to, while `servicerAddress` is the on-chain address (the one related the servicer priv/pub key)

Bad variable naming strikes again, will fix this.

File: relayer/proxy/proxy.go
Comment (red-0ne) [reply relayer/proxy/proxy.go:99]: It should definitely work. Having the whole `Servicer` proto will certainly be better than this in many ways (dev familiarity with the struct and reducing variable naming overhead)

File: relayer/proxy/websockets.go
Comment (red-0ne) [reply relayer/proxy/websockets.go:65]: Yes it is. I'll leave a comment but plan to re-enable it once I get the pubkeys to verify signatures.

File: utils/observable.go
Comment (red-0ne) [reply utils/observable.go:60]: I wish I saw this comment before the previous ones :laughing:

File: smartclient/relayhandler/wshandler.go
Comment (red-0ne) [reply smartclient/relayhandler/wshandler.go:1]: Wise! We should not have a foundation that is in terms of websockets complexity.

File: smartclient/relayhandler/relayhandler.go
Comment (red-0ne) [reply smartclient/relayhandler/relayhandler.go:143]: Yes, it will greatly simplify code that is using it. A first attempt was to extract some of the logic into `session_connection.go`.  But itself should be written in terms of `SessionCache`

**Review by Olshansk — CHANGES_REQUESTED** [review]
@red-0ne Hold off on merging this in just yet. Want to do another look and maybe rename a few things + add more TODOs.

**Comment by Olshansk** [comment]
@red-0ne Wdyt is still remaining for us to merge this work into main?

**Review by Olshansk — APPROVED** [review]
Given that we won't be doing a deep review on this in the alpha repo, I think we should merge this in as is.

**Comment by red-0ne** [comment]
Closing this one in favor of #43 which contains these changes and ring signature capability.

**Pull request closed without merging**
